### PR TITLE
DXS-76

### PR DESCRIPTION
**Pull request by EvansPie**
DXS-76

# What this PR does

- Remove network client token validation on certain cases
- Fix image caching
- Modify order of certain API requests

### CODE REVIEW

File: Example/Podfile [hunk 1]
```diff
@@ -4,8 +4,8 @@ platform :ios, '10.0'
 
 target 'PrimerSDK_Example' do
   pod 'PrimerSDK', :path => '../'
-  pod 'Primer3DS'
-  pod 'PrimerKlarnaSDK'
+#  pod 'Primer3DS'
+#  pod 'PrimerKlarnaSDK'
```
Comment (d4r1091) [inline Example/Podfile:8]: We need to keep these as they are needed to perform BrowserStack tests.
We could think of starting to ignore files - I'll tell you something I've got in mind later today.
```suggestion
pod 'Primer3DS'
pod 'PrimerKlarnaSDK'
```

File: Sources/PrimerSDK/Classes/Core/Checkout Components/PrimerHeadlessUniversalCheckoutUIManager.swift [hunk 2]
```diff
@@ -333,10 +333,10 @@ extension PrimerHeadlessUniversalCheckout {
                             
                         case .continueWithNewClientToken(let newClientToken):
                             firstly {
-                                ClientTokenService.storeClientToken(newClientToken)
+                                ClientTokenService.storeClientToken(newClientToken, isAPIValidationEnabled: true)
                             }
                             .then { () -> Promise<Void> in
-                                let configurationService: PrimerAPIConfigurationServiceProtocol = PrimerAPIConfigurationService(requestDisplayMetadata: true)
```
Comment (d4r1091) [inline Sources/PrimerSDK/Classes/Core/Checkout Components/PrimerHeadlessUniversalCheckoutUIManager.swift:339]: does this change impact the client validation?

File: Sources/PrimerSDK/Classes/Core/Checkout Components/PrimerRawDataManager.swift [hunk 3]
```diff
@@ -239,10 +239,10 @@ extension PrimerHeadlessUniversalCheckout {
                             
                         case .continueWithNewClientToken(let newClientToken):
                             firstly {
-                                ClientTokenService.storeClientToken(newClientToken)
+                                ClientTokenService.storeClientToken(newClientToken, isAPIValidationEnabled: true)
                             }
                             .then { () -> Promise<Void> in
-                                let configurationService = PrimerAPIConfigurationService(requestDisplayMetadata: true)
```
Comment (d4r1091) [inline Sources/PrimerSDK/Classes/Core/Checkout Components/PrimerRawDataManager.swift:245]: Noticing it here as well. I'm sure it does then 👍

File: Sources/PrimerSDK/Classes/Data Models/PrimerConfiguration.swift [hunk 4]
```diff
@@ -87,6 +87,12 @@ extension Response.Body {
             return AppState.current.apiConfiguration?.paymentMethods
         }
         
+        var hasSurchargeEnabled: Bool {
+            let pmSurcharge = AppState.current.apiConfiguration?.clientSession?.paymentMethod?.options?.first(where: { $0["surcharge"] as? Int != nil })
+            let cardSurcharge = AppState.current.apiConfiguration?.clientSession?.paymentMethod?.options?.first(where: { /*$0["type"] as? String == "PAYMENT_CARD" &&*/ (($0["networks"] as? [[String: Any]])?.first(where: { $0["surcharge"] as? Int != nil })) != nil  })
```
Comment (d4r1091) [inline Sources/PrimerSDK/Classes/Data Models/PrimerConfiguration.swift:92]: do we need the commented code here?
asking as the constant name is `let cardSurcharge`

File: Sources/PrimerSDK/Classes/Extensions & Utilities/TimerExtension.swift [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+//  Created by Evangelos on 20/9/22.
+//
+
+#if canImport(UIKit)
+
+import Foundation
+
+internal extension Timer {
+    
+    static func delay(_ timeInterval: TimeInterval) -> Promise<Void> {
```
Comment (d4r1091) [inline Sources/PrimerSDK/Classes/Extensions & Utilities/TimerExtension.swift:14]: 👌

File: Sources/PrimerSDK/Classes/Core/Payment Services/ClientTokenService.swift [hunk 6]
```diff
@@ -7,8 +7,8 @@ internal typealias RawJWTToken = String
 
 internal protocol ClientTokenServiceProtocol {
     static var decodedClientToken: DecodedClientToken? { get }
-    static func storeClientToken(_ clientToken: String) -> Promise<Void>
-    static func storeClientToken(_ clientToken: String, completion: @escaping (Error?) -> Void)
+    static func storeClientToken(_ clientToken: String, isAPIValidationEnabled: Bool) -> Promise<Void>
+    static func resetClientToken()
```
Comment (d4r1091) [inline Sources/PrimerSDK/Classes/Core/Payment Services/ClientTokenService.swift:11]: Just checking in...
I've seen we added this `resetClientToken()` though I've never seen it implemented throughout the changes

File: Sources/PrimerSDK/Classes/Core/Checkout Components/PrimerHeadlessUniversalCheckoutUIManager.swift (hunk 2 above)
Comment (EvansPie) [reply Sources/PrimerSDK/Classes/Core/Checkout Components/PrimerHeadlessUniversalCheckoutUIManager.swift:339]: No, this only makes the request without requesting the display metadata of the payment methods. The display metadata shouldn't be needed when we do a config API call for a required action client token, we should already have them.

File: Sources/PrimerSDK/Classes/Core/Checkout Components/PrimerRawDataManager.swift (hunk 3 above)
Comment (EvansPie) [reply Sources/PrimerSDK/Classes/Core/Checkout Components/PrimerRawDataManager.swift:245]: Nope, as above. It does make the request faster.

File: Sources/PrimerSDK/Classes/Data Models/PrimerConfiguration.swift (hunk 4 above)
Comment (EvansPie) [reply Sources/PrimerSDK/Classes/Data Models/PrimerConfiguration.swift:92]: Nope, we don't

File: Sources/PrimerSDK/Classes/Core/Payment Services/ClientTokenService.swift (hunk 6 above)
Comment (EvansPie) [reply Sources/PrimerSDK/Classes/Core/Payment Services/ClientTokenService.swift:11]: Because it was already there, just not in the protocol

**Review by d4r1091 — APPROVED** [review]

**Pull request merged**